design/deploy.md

```diff
@@ -0,0 +1,98 @@
+# Deploy
+
+A proposal to add support for hosting service deployment to the Polymer CLI.
+
+## Goals
+- Provide new users with a fast way to get a project hosted online.
+- Provide advanced users with an initial bootstrap for more advanced deployments.
+- Demonstrate the [PRPL](https://www.polymer-project.org/2.0/toolbox/server) pattern where HTTP/2 Server Push is available.
+
+## Non-goals
+- Do not try to support every possible deployment environment and configuration, or even a lot of them. Just enough to get users started.
+
+## Design
+
+- Add an `--environment` flag.
+  - Every command knows that it's being run in the context of this environment.
+  - Under the `default` environment, all commands operate as they do today.
+  - Under other environments:
+     - `build` now outputs the boilerplate needed to deploy the project to a given service, in addition to the traditional app build output.
+     - `serve` runs an appropriate HTTP server that emulates the configured hosting environment, instead of polyserve.
+
+- Add an `environments` section to `polymer.json`.
+  - **name**: The name of this environment.
+  - **type**: The type of hosting environment: `app-engine`, `gh-pages`, etc.
+  - Additional type-specific configuration (examples below).
+  - We could also treat environments as a special case of the existing build concept, and include them in the `build` section.
+
+- Add a `deploy` command.
+  - Pushes to the hosting service for the current environment.
+  - Under the `default` environment, offers to help set up a deploy environment.
+     - Could have an `init`-style menu to choose from the available types.
+  - With the `--open` flag, also opens a browser (same as `serve`).
+  - Print the commands we're running so that users can do it themselves.
+
+## Environments
+
```

Comment:
We should have some rough sketch at least of the Environment API - right now we just have `serve()` and `build()` (link to the existing interface too).

For each environment hook we need to specify when it runs, with what arguments, and how it can effect the process it hooks (add warnings, indicate success/failure, modify options, parse flags, etc). We can leave many of those unspecified for now if they wouldn't be used by the two built-in environments we're talking about here. I just want to make sure we're thinking about how this feature evolves.